addons/account/wizard/account_payment_register.py

```diff
@@ -148,17 +149,70 @@ def _get_batches(self):
         if not lines:
             raise UserError(_("You can't open the register payment wizard without at least one receivable/payable line."))
 
+        # Lines are grouped first on 'partner_id','account_id','currency_id' & 'partner_type' and
+        # then on 'partner_bank_id' & 'payment_type'. 'batches' data structure :
+        # {common_pay_val_key:{
+        #         part_bank_pay_type_key : lines,
+        #         part_bank_pay_type_key : lines,
+        #  }}
         batches = {}
+
         for line in lines:
             batch_key = self._get_line_batch_key(line)
-
-            serialized_key = '-'.join(str(v) for v in batch_key.values())
-            batches.setdefault(serialized_key, {
-                'key_values': batch_key,
-                'lines': self.env['account.move.line'],
+            common_pay_val_key = frozendict(batch_key)
+            part_bank_pay_type_key = frozendict({
+                'payment_type': 'inbound' if line.balance > 0.0 else 'outbound',
+                'partner_bank_id': line.move_id.partner_bank_id.id,
             })
-            batches[serialized_key]['lines'] += line
-        return list(batches.values())
+            batches.setdefault(common_pay_val_key, {})
+            batches[common_pay_val_key].setdefault(part_bank_pay_type_key, self.env['account.move.line'])
+            batches[common_pay_val_key][part_bank_pay_type_key] += line
+
+        res = []
+
+        for common_pay_val_key, common_params_group in batches.items():
+            # Group all lines with same 'partner_id','account_id','currency_id' & 'partner_type' in one batch if:
+            # - the resulting batch payment is outbound and there is max 1 partner_bank_id for all outbound lines
+            # or
+            # - the resulting batch payment is inbound and there is max 1 partner_bank_id for all inbound lines
+
+            total = 0.0
+            for lines in common_params_group.values():
+                total += sum(lines.mapped('balance'))
```

Comment:
Compute the aggregation of lines here instead of both here and below.
```python
all_batch_aml = self.env['account.move.line'].browse(
    id 
    for lines in common_params_group.values()
    for id in lines.ids
)
total = sum(all_batch_aml.mapped('balance'))
if float_compare(total, 0.0, precision_digits=line.move_id.currency_id.decimal_places) > 0.0:
    if_grouped_payment_type = 'inbound'
else:
    if_grouped_payment_type = 'outbound'
partner_bank_id_candidates = {
    key['partner_bank_id']
    for key in common_params_group
    if key['partner_bank_id'] and key['payment_type'] == if_grouped_payment_type
}
```
You can keep your way of computing `partner_bank_id_candidates`, but I find it more readable to have one declaration, and never edit the variable later.